lib/tetris.js

```diff
@@ -1,65 +1,102 @@
+  rotate: rotate,
+  clearMatrix: clearMatrix,
+  removeCompleteLines: removeCompleteLines,
+  calculatePointsForLineCompletion: calculatePointsForLineCompletion,
+  checkLineComplete: checkLineComplete,
+  createNewPiece: createNewPiece,
+  move: move,
+  collision: collision
+}
+function initalState (maxWidth, maxHeight) {
+  return {
+    continue: true,
+    player: {
+      matrix: clearMatrix(maxWidth, maxHeight),
+      points: 0,
+      piecePos: {x: 0, y: 5},
+      piece: randomPiece(),
+      pieceRotation: 0
+    },
+    timeControl: {
+      dropInterval: 200,
+      lastDrop: 0
     }
-    return m
-  },
-
-  move: function (originalPlayer, direction) {
-    return move(originalPlayer, newPos(originalPlayer.piecePos, direction))
-  },
-
-  collision: function (player, direction, remove = true) {
-    let matrix = player.matrix
-    if (remove) {
-      matrix = removePiece(player.matrix, player.piece, player.piecePos, player.pieceRotation)
+  }
+}
+function clearMatrix (width = 12, height = 20) {
+  return Array.from(Array(height), () => Array.from(Array(width), () => 0))
+}
+function userEvent (state, eventCode) {
+  if (eventCode === 'ArrowUp') {
+    state.player = rotate(state.player)
+  } else {
+    let moveDirection = {ArrowLeft: 'L', ArrowRight: 'R', ArrowDown: 'D'}[eventCode]
+    if (!collision(state.player, moveDirection)) {
+      state.player = move(state.player, moveDirection)
     }
-    let collide = false
-    runFnAtPiece(matrix, player.piece, newPos(player.piecePos, direction), player.pieceRotation, (matrix, x, y) => { collide = collide || matrix[x] === undefined || matrix[x][y] === undefined || matrix[x][y] })
-    return !!collide
-  },
-
-  rotate: function (originalPlayer) {
-    let player = copy(originalPlayer)
-    player.pieceRotation = player.pieceRotation === 3 ? 0 : player.pieceRotation + 1
-    player.matrix = removePiece(player.matrix, player.piece, player.piecePos, originalPlayer.pieceRotation)
-    player.matrix = insertPiece(player.matrix, player.piece, player.piecePos, player.pieceRotation)
-    return player
   }
-
+  return state
+}
+function rotate (originalPlayer) {
+  let player = copy(originalPlayer)
+  player.pieceRotation = player.pieceRotation === 3 ? 0 : player.pieceRotation + 1
+  player.matrix = removePiece(player.matrix, player.piece, player.piecePos, originalPlayer.pieceRotation)
+  player.matrix = insertPiece(player.matrix, player.piece, player.piecePos, player.pieceRotation)
+  return player
+}
+function tick (state, time) {
+  if (time - state.timeControl.lastDrop <= state.timeControl.dropInterval) {
+    return state
+  }
+  state.timeControl.lastDrop = time
+  if (collision(state.player, 'D')) {
+    if (checkLineComplete(state.player.matrix)) {
+      let roundPoints = calculatePointsForLineCompletion(state.player.matrix)
+      if (roundPoints === undefined) {
+        // Cheating Detected
+        console.log('roundPoints undefined')
+        return false
```

Comment:
this function return state, so it should set state.continue to false and return state